### PR TITLE
Allow injecting authentication components for ThriftMetastore

### PR DESCRIPTION
**Pull request by phd3**
## Description

We've a requirement for using an external service to authenticate to HMS as Trino, and also fetch HMS delegation tokens for the end-user. So proposing this change to make these components to be provided by an external module.  ( Using optional binder to avoid having to make ThriftMetastoreAuthenticationType enum a string)

The PR contains following major changes:

* Rename the Thrift metastore client factories to distinguish them better given what the interface exposes

| Interface old name                          | Signature                                                     | Return Type           | New name                                 |
|---------------------------------------------|---------------------------------------------------------------|-----------------------|------------------------------------------|
| ThriftMetastoreFactory (highest level)      | createMetastore(Optional\<ConnectorIdentity\> identity)         | ThriftMetastore       | ThriftMetastoreFactory (unchanged)       |
| TokenDelegationThriftMetastoreFactory       | createMetastoreClient(Optional\<ConnectorIdentity\> identity)   | ThriftMetastoreClient | **IdentityAwareMetastoreClientFactory**      |
| MetastoreLocator                            | createMetastoreClient(Optional\<String\> token)         | ThriftMetastoreClient | **TokenAwareMetastoreClientFactory**          |
| ThriftMetastoreClientFactory (lowest level) | create(HostAndPort address, Optional\<String\> delegationToken) | ThriftMetastoreClient | ThriftMetastoreClientFactory (unchanged) |

* Narrow down usage of `ThriftMetastoreAuthenticationType` to make auth decision. (Also makes `TokenDelegationThriftMetastoreFactory` aka `IdentityAwareMetastoreClientFactory` an interface)

* Allow binding your own thrift authentication module 

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by phd3** [comment]
@dain @electrum can you please take a look?

**Comment by phd3** [comment]
Rebased with some modifications to align with the recent refactoring. some more details in PR description. @dain @losipiuk does this approach look reasonable?

**Comment by phd3** [comment]
cc @electrum

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftMetastoreAuthenticationModule.java [hunk 1]
```diff
@@ -35,29 +36,28 @@
...
         switch (type) {
             case NONE:
-                return new NoHiveMetastoreAuthenticationModule();
+                // already bound optionally
+                break;
             case KERBEROS:
-                return new KerberosHiveMetastoreAuthenticationModule();
+                install(new KerberosHiveMetastoreAuthenticationModule());
         }
-        throw new AssertionError("Unknown authentication type: " + type);
```
Comment (dain) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftMetastoreAuthenticationModule.java:50]: Is the goal to allow additional types in the future?  If so, I'd rewrite this as:
```java
if (buildConfigObject(ThriftMetastoreAuthenticationConfig.class).getAuthenticationType() == KERBEROS) {
    install(new KerberosHiveMetastoreAuthenticationModule());
}

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftMetastoreAuthenticationModule.java [hunk 2]
```diff
@@ -35,29 +36,28 @@
...
-    public static class NoHiveMetastoreAuthenticationModule
+    public static class OptionalDefaultHiveMetastoreAuthenticationModule
             implements Module
     {
         @Override
         public void configure(Binder binder)
         {
-            binder.bind(IdentityAwareMetastoreClientFactory.class).to(UgiBasedMetastoreClientFactory.class).in(SINGLETON);
-            binder.bind(HiveMetastoreAuthentication.class).to(NoHiveMetastoreAuthentication.class).in(SINGLETON);
+            newOptionalBinder(binder, IdentityAwareMetastoreClientFactory.class)
```
Comment (dain) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftMetastoreAuthenticationModule.java:57]: Inline these into the sole caller above, and remove this class

**Review by dain — COMMENTED** [review]
Couple of minor comments, but otherwise looks good

**Review by dain — APPROVED** [review]

**Pull request merged**